// @stimulus/core/dist/module/src/context_set.js
import { Context } from "./context";
var ContextSet = /** @class */ (function () {
    function ContextSet(router, identifier, controllerConstructor) {
        this.router = router;
        this.identifier = identifier;
        this.controllerConstructor = controllerConstructor;
        this.contextsByElement = new WeakMap;
        this.connectedContexts = new Set;
    }
    Object.defineProperty(ContextSet.prototype, "application", {
        get: function () {
            return this.router.application;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextSet.prototype, "contexts", {
        get: function () {
            return Array.from(this.connectedContexts);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(ContextSet.prototype, "size", {
        get: function () {
            return this.connectedContexts.size;
        },
        enumerable: true,
        configurable: true
    });
    ContextSet.prototype.connect = function (element) {
        var context = this.fetchContextForElement(element);
        if (context && !this.connectedContexts.has(context)) {
            this.connectedContexts.add(context);
            context.connect();
        }
    };
    ContextSet.prototype.disconnect = function (element) {
        var context = this.fetchContextForElement(element);
        if (context && this.connectedContexts.has(context)) {
            this.connectedContexts.delete(context);
            context.disconnect();
        }
    };
    ContextSet.prototype.getContextForElement = function (element) {
        return this.contextsByElement.get(element);
    };
    ContextSet.prototype.fetchContextForElement = function (element) {
        var context = this.contextsByElement.get(element);
        if (!context) {
            context = new Context(this, element);
            this.contextsByElement.set(element, context);
        }
        return context;
    };
    return ContextSet;
}());
export { ContextSet };
